### PR TITLE
enum checks

### PR DESCRIPTION
**Pull request by matevz**
PR for #2527:
- adds `KindInvalid` to `RuntimeKind` enum,
- adds `KindInvalid` to `CommitteeKind` enum,
- adds `ContextInvalid` to `ContextType` enum,
- adds `HeaderInvalid` to `HeaderType` enum,
- adds `Invalid` to `artifactType` enum both in go and rust,
- in `CommitteeKind` enum renames `KindExecutor` -> `KindComputeExecutor`, `KindTransactionScheduler` -> `KindComputeTxnScheduler`, and `KindMerge` -> `KindComputeMerge`.

### CODE REVIEW

File: go/runtime/transaction/transaction.go [hunk 1]
```diff
@@ -68,7 +77,8 @@ func (ak *artifactKind) UnmarshalBinary(data []byte) error {
 
 var (
 	// txnKeyFmt is the key format used for transaction artifacts.
-	txnKeyFmt = keyformat.New('T', &hash.Hash{}, artifactKind(0))
+	// TODO Matevz: What exactly does third argument represent? I wanted to use kindInvalid (as default?), but I'm getting crashes in transaction_test.go then.
```
Comment (kostko) [inline go/runtime/transaction/transaction.go]: You can check the `keyformat` docs. The first argument is the type/tag while all following arguments specify sizes of key atoms. Only the type matters, not the value, so `artifactKind(0)` is correct. This is NOT a default value.

File: go/scheduler/api/api.go [hunk 2]
```diff
@@ -67,25 +67,30 @@ type CommitteeNode struct {
 type CommitteeKind uint8
 
 const (
+	// KindCompute is a compute committee.
+	KindInvalid CommitteeKind = 0
+
 	// KindExecutor is an executor committee.
-	KindExecutor CommitteeKind = 0
+	KindExecutor CommitteeKind = 1
```
Comment (kostko) [inline go/scheduler/api/api.go]: While we are here, can you refactor this to be:
* `KindComputeExecutor`
* `KindComputeTxnScheduler`
* `KindComputeMerge`

And have these be grouped together in value space.

File: go/scheduler/api/api.go [hunk 3]
```diff
@@ -67,25 +67,30 @@ type CommitteeNode struct {
...
 	// MaxCommitteeKind is a dummy value used for iterating all committee kinds.
-	MaxCommitteeKind = 4
+	MaxCommitteeKind = 5
 )
 
 // NeedsLeader returns if committee kind needs leader role.
 func (k CommitteeKind) NeedsLeader() bool {
 	switch k {
+	case KindInvalid:
+		panic(fmt.Errorf("scheduler: KindInvalid should not be queried for a leader"))
```
Comment (kostko) [inline go/scheduler/api/api.go]: Panicking here seems dangerous, what if an invalid kind is received over the wire?

File: go/runtime/transaction/transaction.go (hunk 1 above)
Comment (matevz) [reply go/runtime/transaction/transaction.go]: Marshallable `artifactKind` is required in order to obtain its size in bytes. I changed this to first valid value now, i.e. `artifactKind(1)`.

File: go/scheduler/api/api.go [hunk 4]
```diff
@@ -67,46 +67,53 @@ type CommitteeNode struct {
...
 // NeedsLeader returns if committee kind needs leader role.
-func (k CommitteeKind) NeedsLeader() bool {
+func (k CommitteeKind) NeedsLeader() (bool, error) {
 	switch k {
-	case KindExecutor:
-		return false
-	case KindMerge:
-		return false
+	case KindInvalid:
+		return false, fmt.Errorf("scheduler: KindInvalid should not be queried for a leader")
```
Comment (kostko) [inline go/scheduler/api/api.go]: What about all the other kinds that are also invalid (e.g., greater than or equal to `MaxCommitteeKind`)?

File: runtime/src/transaction/tree.rs [hunk 5]
```diff
@@ -17,13 +17,17 @@ use crate::{
 // NOTE: This should be kept in sync with go/runtime/transaction/transaction.go.
 
 #[derive(Debug)]
+#[repr(u8)]
 enum ArtifactKind {
-    Input,
-    Output,
+    _Invalid = 0,
```
Comment (kostko) [inline runtime/src/transaction/tree.rs]: Why is this needed? Just remove this invalid kind and you will not be able to construct one accidentally?

File: runtime/src/transaction/tree.rs [hunk 6]
```diff
@@ -17,13 +17,17 @@ use crate::{
...
+    Input = ARTIFACT_KIND_INPUT,
+    Output = ARTIFACT_KIND_OUTPUT,
 }
 
-const ARTIFACT_KIND_INPUT: u8 = 0;
-const ARTIFACT_KIND_OUTPUT: u8 = 1;
+// Workaround because rust doesn't support `as u8` inside match arms.
+// See https://github.com/rust-lang/rust/issues/44266
+const ARTIFACT_KIND_INPUT: u8 = 1;
```
Comment (kostko) [inline runtime/src/transaction/tree.rs]: At least use `ArtifactKind::Input as u8` here (the workaround is only for match arms).

File: go/oasis-node/cmd/registry/runtime/runtime.go [hunk 7]
```diff
@@ -274,6 +274,10 @@ func runtimeFromFlags() (*registry.Runtime, signature.Signer, error) {
 			)
 			return nil, nil, fmt.Errorf("invalid runtime flags")
 		}
+	case registry.KindInvalid:
+		{
```
[the remark this answers is not in the document]
Comment (matevz) [reply go/oasis-node/cmd/registry/runtime/runtime.go]: Meh, too much c++ over the weekend :wink:

File: go/runtime/transaction/transaction.go [hunk 8]
```diff
@@ -254,7 +268,9 @@ func (t *Tree) GetTransactions(ctx context.Context) ([]*Transaction, error) {
 	curTx.Empty()
 
 	var txs []*Transaction
+	i := 0
 	for it.Seek(txnKeyFmt.Encode()); it.Valid(); it.Next() {
+		i++
```
[the remark this answers is not in the document]
Comment (matevz) [reply go/runtime/transaction/transaction.go]: Debugging leftovers, sorry.

File: go/roothash/api/commitment/pool_test.go [hunk 9]
```diff
@@ -204,6 +205,7 @@ func TestPoolSingleCommitmentTEE(t *testing.T) {
 
 	rt := &registry.Runtime{
 		ID:          rtID,
+		Kind:        registry.KindInvalid,
```
Comment (kostko) [inline go/roothash/api/commitment/pool_test.go]: You fixed the one above, but all the others should also be changed to `KindCompute`.

Comment (matevz) [reply go/roothash/api/commitment/pool_test.go]: fixed
